refs #25285 - deprecate @repos variable

stbenjam opened the pull request:
During the addition of additional media support, `@repos` was erroneously
removed instead of being deprecated.  I have the deprecation set to 1.21.

If #25285 will not be accepted for 1.20, I'll change it to be 1.22.

stbenjam committed:
refs #25285 - deprecate @repos variable

During the addition of additional media support, @repos was erronenously
removed instead of being deprecated.

lzap commented:
Failures on CI. Many.

stbenjam commented:
Yea basically everything that renders something. Any ideas how to deprecate this without causing all the tests to fail? It'd have to be pretty hacky to detect when a variable is referenced before calling the deprecation notification.

sean797 commented on app/models/operatingsystem.rb:
```diff
@@ -119,6 +119,23 @@ def self.families_as_collection
+  #    :gpgcheck => 1
+  #  }]
+  def repos(host)
+    Foreman::Deprecation.deprecation_warning('1.21', 'Additional repositories should be provided through additional media features of medium providers instead.')
```
Mind also telling people what is actually deprecated in this message?

sean797 reviewed (commented):
@stbenjam add it to `config/as_deprecation_whitelist.yaml`

1.21 isn't long away, is there a reason to not to remove this in 1.22 instead?

stbenjam commented:
As far as 1.21 or 1.22, I don't think it matters, as there's no evidence anyone has ever used this, but I updated the message and moved the deprecation to 1.22.

stbenjam commented:
[test upgrade]

stbenjam commented:
This is ready for another look.

lzap commented on lib/foreman/renderer/scope/variables/base.rb:
```diff
@@ -75,6 +75,7 @@ def kickstart_attributes
             @arch      = architecture_name
             @osver     = major.try(:to_i)
             @mediapath = mediumpath(@medium_provider) if @medium_provider
+            @repos     = repos(host)
```
This will put the deprecation warning into logs every time we render a provision template. Not sure if this is what we want. Opinions? I suggest to comment the warning line out but keep it in the codebase including the YAML file so we are notified in the future to remove it.

stbenjam replied:
The only way a user would know then would be the release notes, since the logs won't tell them.  I think that's fine since I really doubt anyone uses this.  But, since @tbrisker asked for the deprecation maybe he has an opinion about it.

tbrisker replied:
I agree that having the warning every time a template is rendered is too much noise. I'm trying to figure out if there is some way of only logging that when the variable is accessed, but so far no luck. Perhaps we could show the deprecation only if the function isn't called from here? i.e. `Foreman::Deprecation ... unless caller.any? {|c| c.include? 'scope/variables/base'}`

lzap replied:
I am for comment only, this is not widely used variable I think.

stbenjam replied:
Updated to comment only.  Left the item in the whitelist, so we remember to remove it in 1.22. I don't think firing the message based only on other callers is helpful, if anyone uses this it'd be through the `@repos` variable.

tbrisker replied:
OK, i'm fine with this. Thanks @stbenjam and @lzap !
